elastic/logstash
# Fix/introduce explicit dependencies

## Release notes

[rn:skip]

## What does this PR do?


This is a preparatory PR to switch to Gradle 7. 
Gradle 7 introduced a check for [implicit task dependencies](https://docs.gradle.org/7.2/userguide/validation_problems.html#implicit_dependency). 
In particular the `copyPluginAlias` task contains the property assignment:
```
destinationDir = projectDir
```
and `destinationDir` is marked with `@OutputDirectory` so every task that read from that output has an implicit dependency on this task. In this configuartion the `destinationDir` is assigned to `projectDir` so it means that every task that read from `projectDir` produces a warn. Many tasks in Gradle read and write to the `projectDir`.
This PR rework the configuration of Copy task to avoid copying same file in multiple location (from which comes the dependency to `buildDir`) in favor of more fine grained tasks that reduces output directory width.


## Why is it important/What is the impact to the user?


User isn't affected by this

## Checklist




## Author's Checklist



## How to test this PR locally


- install a system gradle 7 (`sdk install gradle 7.2`) on your system
- if not set as default, select it (`sdk use gradle 7.2`)
- run any task on `master` branch, `gradle clean`, *not* `./gradlew`
- run it on the branch and check it's ok

## Related issues


- Relates #13177 



## Logs


This is a sample of an error by Gradle 7
```
> Task :jar
Execution optimizations have been disabled for task ':jar' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: '/home/andrea/workspace/logstash_andsel/build/classes/java/main'. Reason: Task ':jar' uses this output of task ':copyPluginTestAlias' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.1.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/home/andrea/workspace/logstash_andsel/build/resources/main'. Reason: Task ':jar' uses this output of task ':copyPluginTestAlias' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.1.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/home/andrea/workspace/logstash_andsel/build/tmp/jar/MANIFEST.MF'. Reason: Task ':jar' uses this output of task ':copyPluginTestAlias' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.1.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
```
